pallets/loans/src/functions.rs

```diff
@@ -125,68 +125,99 @@ impl<T: Config> Pallet<T> {
 		pool_id: PoolIdOf<T>,
 		loan_id: T::LoanId,
 		interest_rate_per_sec: T::Rate,
 		loan_type: LoanType<T::Rate, T::Balance>,
 	) -> Result<u32, DispatchError> {
-		Loan::<T>::try_mutate(pool_id, loan_id, |loan| -> Result<u32, DispatchError> {
-			let loan = loan.as_mut().ok_or(Error::<T>::MissingLoan)?;
+		let now = Self::now();
+		ensure!(loan_type.is_valid(now), Error::<T>::LoanValueInvalid);
 
-			// ensure loan is created or priced but not yet borrowed against
-			ensure!(
-				loan.status == LoanStatus::Created || loan.status == LoanStatus::Active,
-				Error::<T>::LoanIsActive
-			);
-			let mut active_loans = ActiveLoans::<T>::get(pool_id);
-			if loan.status == LoanStatus::Active {
-				if let Some((idx, active_loan)) = active_loans
-					.iter()
-					.enumerate()
-					.find(|(_, active_loan)| active_loan.loan_id == loan_id)
-				{
-					ensure!(
-						active_loan.total_borrowed == Zero::zero(),
-						Error::<T>::LoanIsActive
-					);
-					active_loans.remove(idx);
-				}
-			}
+		ensure!(
+			interest_rate_per_sec >= One::one(),
+			Error::<T>::LoanValueInvalid
+		);
 
-			// ensure loan_type is valid
-			let now = Self::now();
-			ensure!(loan_type.is_valid(now), Error::<T>::LoanValueInvalid);
+		let active_loan = PricedLoanDetails {
+			loan_id,
+			loan_type,
+			interest_rate_per_sec,
+			origination_date: None,
+			normalized_debt: Zero::zero(),
+			total_borrowed: Zero::zero(),
+			total_repaid: Zero::zero(),
+			write_off_status: WriteOffStatus::None,
+			last_updated: now,
+		};
 
-			// ensure interest_rate_per_sec >= one
-			ensure!(
-				interest_rate_per_sec >= One::one(),
-				Error::<T>::LoanValueInvalid
-			);
+		let mut active_loans = ActiveLoans::<T>::get(pool_id);
+		active_loans
+			.try_push(active_loan)
+			.map_err(|_| Error::<T>::TooManyActiveLoans)?;
+		let count = active_loans.len();
+		ActiveLoans::<T>::insert(pool_id, active_loans);
 
-			let active_loan = PricedLoanDetails {
-				loan_id,
-				loan_type,
-				interest_rate_per_sec,
-				origination_date: None,
-				normalized_debt: Zero::zero(),
-				total_borrowed: Zero::zero(),
-				total_repaid: Zero::zero(),
-				write_off_status: WriteOffStatus::None,
-				last_updated: Self::now(),
-			};
-
-			active_loans
-				.try_push(active_loan)
-				.map_err(|_| Error::<T>::TooManyActiveLoans)?;
-			let count = active_loans.len();
-			ActiveLoans::<T>::insert(pool_id, active_loans);
-
-			// update the loan status
-			loan.status = LoanStatus::Active;
-
-			Ok(count.try_into().unwrap())
-		})
+		Ok(count.try_into().unwrap())
+	}
+
+	pub(crate) fn price_active_loan(
+		pool_id: PoolIdOf<T>,
+		loan_id: T::LoanId,
+		interest_rate_per_sec: T::Rate,
+		loan_type: LoanType<T::Rate, T::Balance>,
+	) -> Result<u32, DispatchError> {
+		let now = Self::now();
+		ensure!(loan_type.is_valid(now), Error::<T>::LoanValueInvalid);
+
+		ensure!(
+			interest_rate_per_sec >= One::one(),
+			Error::<T>::LoanValueInvalid
+		);
+
+		Self::try_mutate_active_loan(
```

Comment:
In the issue description, it's not clear to me if we need to ensure that the current debt should be > 0 or if it should work for 0 too. The current implementation supposes the second option. Please check this 😝

Reply:
Yeah the second option is right 👍